versions/version3/version3.js

```diff
@@ -311,36 +338,51 @@ class DigitalRain {
         );
       };
 
-      const textTweak = (context, character, xloc, yloc, charPos) => {
+      const textTweak = (context, character, xloc, yloc, charPos, color) => {
+        const originalStyle = context.fillStyle;
+        if (color != null) {
+          context.fillStyle = color;
+        }
         context.fillText(
           character,
           Math.floor(xloc),
           Math.floor(yloc - this.fontSize * charPos),
           rectTrim
         );
-      };
-
-      const getRandomChar = () => {
-        const randNum = Math.floor(Math.random() * this.characters.length);
-        return this.characters.charAt(randNum);
+        context.fillStyle = originalStyle;
       };
 
       /*
       Think of the end of the stream being at the top of the code here...
     */
 
       // Clean up at the end of the stream
-      this.ctx2.fillStyle = "black";
-      fillRectTweak(this.ctx2, stream.XLOC, stream.YLOC, stream.streamLength);
+      if (stream.characters.length > stream.streamLength) {
+        stream.characters.pop();
+      }
+      // this.ctx2.fillStyle = "black";
+      // fillRectTweak(this.ctx2, stream.XLOC, stream.YLOC, stream.streamLength);
       this.ctx.fillStyle = "black";
       fillRectTweak(this.ctx, stream.XLOC, stream.YLOC, stream.streamLength);
 
       // Shading with rgba() is detrimental to performance because
       // the browser has to calculate the blending on each draw
       if (this.shading) {
-        for (let i = 4; i < stream.streamLength - 3; i++) {
-          this.ctx2.fillStyle = "rgba(0, 0, 0, 0.075)";
-          fillRectTweak(this.ctx2, stream.XLOC, stream.YLOC, i);
+        for (let i = 4; i < stream.characters.length; i++) {
+          const char = stream.characters[i];
+          const settledRgb = parseHex(this.settledColor);
+          const multiplier = 1 - (i + 1) / stream.streamLength;
+          const shadedRgb = multiplyRgb(settledRgb, multiplier);
+          const shadedColor = rgbToHex(shadedRgb);
+          clearRectTweak(this.ctx, stream.XLOC, stream.YLOC, i + 1);
+          textTweak(
+            this.ctx,
+            char,
+            stream.XLOC,
+            stream.YLOC,
+            i + 1,
+            shadedColor
+          );
```

Comment:
Alternative way of implementing the gradient shadow effect. This should be more performant because this approach does not use alpha channel.